src/compute-client/src/controller.rs

```diff
@@ -889,19 +898,219 @@ impl<T: ComputeControllerTimestamp> CollectionState<T> {
+        initial_as_of: Option<Antichain<T>>,
+        refresh_schedule: Option<RefreshSchedule>,
+        upper: Antichain<T>,
+    ) -> CollectionIntrospection<T> {
+        let self_ = CollectionIntrospection {
+            collection_id,
+            introspection_tx,
+            refresh_introspection_state: refresh_schedule.map(|refresh_schedule| {
+                if let Some(initial_as_of) = initial_as_of {
+                    Some(RefreshIntrospectionState::new(
+                        refresh_schedule,
+                        initial_as_of,
+                        upper,
+                    ))
+                } else {
+                    soft_panic_or_log!("if we have a refresh_schedule, then it's an MV, so we should also have an initial_as_of");
+                    None
+                }
+            }).flatten()
+        };
+
+        self_
+            .refresh_introspection_state
+            .as_ref()
+            .map(|refresh_introspection_state| {
+                let insertion = refresh_introspection_state.row_for_collection(collection_id);
+                self_.send(
+                    IntrospectionType::ComputeMaterializedViewRefreshes,
+                    vec![(insertion, 1)],
+                );
+            });
+
+        self_
+    }
+
+    pub fn frontier_update(&mut self, write_frontier: &Antichain<T>) {
+        if let Some(refresh_introspection_state) = &mut self.refresh_introspection_state {
+            // Old row to retract based on old state.
+            let retraction = refresh_introspection_state.row_for_collection(self.collection_id);
+            // Update state.
+            refresh_introspection_state.frontier_update(write_frontier);
+            // New row to insert based on new state.
+            let insertion = refresh_introspection_state.row_for_collection(self.collection_id);
+            // Send changes.
+            self.send(
+                IntrospectionType::ComputeMaterializedViewRefreshes,
+                vec![(retraction, -1), (insertion, 1)],
+            );
+        }
+    }
+}
+
+impl<T> Drop for CollectionIntrospection<T> {
+    fn drop(&mut self) {
+        self.refresh_introspection_state
+            .as_ref()
+            .map(|refresh_introspection_state| {
+                let retraction = refresh_introspection_state.row_for_collection(self.collection_id);
+                self.send(
+                    IntrospectionType::ComputeMaterializedViewRefreshes,
+                    vec![(retraction, -1)],
+                );
+            });
+    }
+}
+
+/// Information needed to compute introspection updates for a REFRESH materialized view when the
+/// write frontier advances.
+#[derive(Debug)]
+struct RefreshIntrospectionState<T> {
+    // Immutable properties of the MV
+    refresh_schedule: RefreshSchedule,
+    initial_as_of: Antichain<T>,
+    // Refresh state
+    next_refresh: Datum<'static>,
+    last_completed_refresh: Datum<'static>,
```

Comment:
I tried to do it that way originally, but then the problem is that the
```
impl<T> Drop for CollectionIntrospection<T>
```
would need to become
```
impl<T: ComputeControllerTimestamp> Drop for CollectionIntrospection<T>
```
which is bad, because then all the containing structs would also need `<T: ComputeControllerTimestamp>` (as opposed to having `T: ComputeControllerTimestamp` on just certain impls). An old commit that shows this: 93ff009ab815a220a58194877d9bf77767459749

Reply:
Yup, that'll happen, but I think it's OK because we're trading off more readable code with a trait bound. Anyways, no strong feeling to changing it as part of this PR.

Reply:
I've added comments on the Datum fields to indicate what kind of Datum can come there:
```
    // Refresh state
    next_refresh: Datum<'static>, // Null or an MzTimestamp
    last_completed_refresh: Datum<'static>, // Null or an MzTimestamp
```